docs/proposals/component-registration.md

```diff
@@ -0,0 +1,337 @@
+
+Note: The endpoint should contain cause where possible, but at least the current condition of the components.
+
+#### Notification
+
+As a k8s operator, I want to be notified when the cluster requires human intervention so that I can intervene.
+
+As an external notification system (e.g. PagerDuty), I want an endpoint that responds to regular requests with the current cluster condition so that I can determine (with my own rule set) when to notify the k8s operator.
+
+Note: What condition requires human intervention is not a rule set that can be hard-coded, is likely to be different for each cloud platform, changes based on component configuration, and is likely to require a time-series database.
+
+#### Reporting
+
+As a k8s operator, I want to see a report/graph of the readiness/uptime/availability of the cluster over time so that I can make/satisfy SLAs.
+
+Note: At least requires a time-series database and probably a reporting engine.
+
+#### Self-Healing
+
+As an k8s operator, I want the cluster to self-heal when individual components crash so that I don't get notified or need to intervene.
+
+As a local process monitoring tool (e.g. monit), I want to be able to determine the condition of the component process I am monitoring to determine if it needs to be restarted.
+
+As a remote container/vm monitoring tool (e.g. bosh monitor), I want to be able to determine the heath of the component container/vm I am monitoring.
+
+Note: Self-healing requires knowing how to deploy the cluster components, which is different for each cloud platform and changes based on component configuration.
+
+
+## Design
+
+### API Changes
+
+Add a single component status endpoint (`/components/status`) that lists the last known condition of all the k8s components.
+
+Add a single component endpoint (`/components`) that lists the spec/status of all the k8s components.
+
+Add individual component endpoints (`/component/status/<name>`) that show the status of each k8s component.
+
+Add a LivenessProbe to the component spec that describes how to probe whether the component is running (including during bootstrapping/startup).
+
+Add a ReadinessProbe to the component spec that describes how to probe whether the component is fully functional and ready to be used.
+
+Group components by type, to allow multiple instances of each type, to enable HA cluster deployments.
+
+Use the existing name generator with component type as the prefix to ensure component names are unique. Return the generated name in the create response.
+
+#### Component API
+
+The following is a draft of how the Component model might look.
+
+```
+// ComponentType is the name of a group of one or more component instances
+type ComponentType string
+
+// ComponentSpec defines a component and how to verify its status
+type ComponentSpec struct {
+	// Type of the component
+	Type ComponentType `json:"type"`
```

Comment:
No port/url information is included, only implicitly in the probes. In other words, this is not a discovery mechanism. I guess that's intentionally and supposed to be solved by using services (like the kubernetes service pointing to the apiserver).